src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,606 @@
+//! # Interpreatation of Miniscript scripts
+//!
+//! Traits and implementations to support for interpretation of miniscripts.
+//!
+//!
+use bitcoin_hashes::sha256;
+use secp256k1;
+
+use Error;
+use miniscript::astelem::AstElem;
+use ToPublicKey;
+use secp256k1::{Secp256k1, Signature, VerifyOnly};
+use bitcoin_hashes::Hash;
+use bitcoin;
+
+///Primitives for Miniscript: a vector of public keys(sigs), hashlocks and timelocks
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Primitives{
+    pub pks: Vec<bitcoin::PublicKey>,
+    pub hashlocks : Vec<Vec<u8>>,
+    pub timelocks : Vec<u32>
+}
+
+impl Primitives
+{
+    fn join(&mut self, other: &Primitives){
+        self.pks.extend(other.pks.clone());
+        self.hashlocks.extend(other.hashlocks.clone());
+        self.timelocks.extend(other.timelocks.clone());
+    }
+}
+
+
+/// Definition of Stack Element of the Witness stack used for interpretation of Miniscript.
+/// All stack elements with vec![] go to Dissatisfied and vec![1] are marked to Satisfied.
+/// Others are directly pushed as witness
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum StackElement {
+    /// Result of a satisfied E/T/W/F
+    Satisfied,
+    /// Result of a dissatisfied E/W
+    Dissatisfied,
+    /// Input from the witness stack
+    Witness(Vec<u8>),
+}
```

Comment:
Can you try
```

#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum StackElement<'a> {
    /// Result of a satisfied E/T/W/F, or input to `IF`/`NOTIF`
    Satisfied,
    /// Result of a dissatisfied E/W, or input to `IF`/`NOTIF`
    Dissatisfied,
    /// Other data from the witness stack
    Push(&'a [u8]),
}
```
You'll run into a bunch of weird lifetime issues doing the `'a` thing - if you have too much trouble let me know and I'll add a commit which does this. But the upside is that you'll avoid a bunch of extra allocations.

But at the very least can you rename `Witness` to `Push` (my, bad!). I didn't realize when I made my original suggestion that `Satisfied` and `Dissatisfied` might also be input from the witness stack, so naming one variant `Witness` is a bit confusing.